tests/scene/test_animation.h

```diff
@@ -0,0 +1,314 @@
+	CHECK(animation->scale_track_interpolate(0, 0.0, nullptr) == ERR_INVALID_PARAMETER);
+	CHECK(Math::is_zero_approx(animation->bezier_track_interpolate(0, 0.0)));
+	CHECK(animation->blend_shape_track_interpolate(0, 0.0, nullptr) == ERR_INVALID_PARAMETER);
+	ERR_PRINT_ON;
+}
+
+TEST_CASE("[Animation] Create 3D position track") {
+	Ref<Animation> animation = memnew(Animation);
+	const int track_index = animation->add_track(Animation::TYPE_POSITION_3D);
+	animation->track_set_path(track_index, NodePath("Enemy:position"));
+	animation->position_track_insert_key(track_index, 0.0, Vector3(0, 1, 2));
+	animation->position_track_insert_key(track_index, 0.5, Vector3(3.5, 4, 5));
+
+	CHECK(animation->get_track_count() == 1);
+	CHECK(!animation->track_is_compressed(0));
+	CHECK(Vector3(animation->track_get_key_value(0, 0)).is_equal_approx(Vector3(0, 1, 2)));
+	CHECK(Vector3(animation->track_get_key_value(0, 1)).is_equal_approx(Vector3(3.5, 4, 5)));
+
+	Vector3 r_interpolation;
+
+	CHECK(animation->position_track_interpolate(0, -0.2, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(0, 1, 2)));
+
+	CHECK(animation->position_track_interpolate(0, 0.0, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(0, 1, 2)));
+
+	CHECK(animation->position_track_interpolate(0, 0.2, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(1.4, 2.2, 3.2)));
+
+	CHECK(animation->position_track_interpolate(0, 0.4, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(2.8, 3.4, 4.4)));
+
+	CHECK(animation->position_track_interpolate(0, 0.5, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(3.5, 4, 5)));
+
+	CHECK(animation->position_track_interpolate(0, 0.6, &r_interpolation) == OK);
+	CHECK(r_interpolation.is_equal_approx(Vector3(3.5, 4, 5)));
+
+	// 3D position tracks always use linear interpolation for performance reasons.
+	CHECK(Math::is_equal_approx(animation->track_get_key_transition(0, 0), real_t(1.0)));
+	CHECK(Math::is_equal_approx(animation->track_get_key_transition(0, 1), real_t(1.0)));
+
+	// This is a 3D position track, so the methods below should return errors.
+	ERR_PRINT_OFF;
+	CHECK(animation->value_track_interpolate(0, 0.0).is_null());
+	CHECK(animation->rotation_track_interpolate(0, 0.0, nullptr) == ERR_INVALID_PARAMETER);
+	CHECK(animation->scale_track_interpolate(0, 0.0, nullptr) == ERR_INVALID_PARAMETER);
+	CHECK(Math::is_zero_approx(animation->bezier_track_interpolate(0, 0.0)));
+	CHECK(animation->blend_shape_track_interpolate(0, 0.0, nullptr) == ERR_INVALID_PARAMETER);
+	ERR_PRINT_ON;
+}
+
+TEST_CASE("[Animation] Create 3D rotation track") {
+	Ref<Animation> animation = memnew(Animation);
+	const int track_index = animation->add_track(Animation::TYPE_ROTATION_3D);
+	animation->track_set_path(track_index, NodePath("Enemy:rotation"));
+	animation->rotation_track_insert_key(track_index, 0.0, Quaternion(Vector3(0, 1, 2)));
```

Comment:
This is odd. Isn't this supposed to go to basis first?

Reply:
In `master`, this creates a Quaternion from an Euler angle.